File: src/test/java/voidchess/figures/FigureTest.java
package voidchess.figures;

import org.testng.annotations.Test;
import voidchess.board.ChessGameInterface;
import voidchess.board.LastMoveProvider;
import voidchess.board.SimpleArrayBoard;
import voidchess.helper.Move;
import voidchess.helper.Position;

import java.util.LinkedList;
import java.util.List;

import static org.mockito.Mockito.mock;
import static org.testng.Assert.*;

/**
 * @author stephan
 */
public class FigureTest {

    @Test
    public void testIsDifferentColor() {
        Figure figure1 = new Queen(true, Position.get(0, 0));
        Figure figure2 = new Rook(true, Position.get(0, 0));
        Figure figure3 = new Bishop(false, Position.get(0, 0));
        assertTrue(figure1.hasDifferentColor(figure3));
        assertFalse(figure2.hasDifferentColor(figure1));
        assertFalse(figure3.hasDifferentColor(figure3));
    }

    @Test
    public void testCanBeHitByEnpasent() {
        assertFalse(
                new Knight(
                        true,
                        Position.get(0, 0)
                ).canBeHitByEnpasent(),
                "should be false for all voidchess.figures but pawns"
        );
    }

    @Test
    public void testToString() {
        Position pos = Position.byCode("g4");
        Figure figure1 = new Bishop(true, pos);
        assertEquals("Bishop-white-g4", figure1.toString());
        Figure figure2 = new Bishop(false, pos);
        assertEquals("Bishop-black-g4", figure2.toString());
    }

    @Test
    public void testSubtypes() {
        Position pos = Position.byCode("g4");
        new Pawn(true, pos);
        new Rook(true, pos);
        new Knight(true, pos);
        new Bishop(true, pos);
        new Queen(true, pos);
        new King(true, pos);
    }

    @Test
    public void testIsWhite() {
        Position pos = Position.byCode("g4");

        Figure figure1 = new Rook(true, pos);
        assertTrue(figure1.isWhite());

        Figure figure2 = new Knight(false, pos);
        assertFalse(figure2.isWhite());
    }

    @Test
    public void testFigureMoved() {
        Position from = Position.byCode("c1");
        Position to = Position.byCode("g5");
        Position thirdpos = Position.byCode("f5");

        Figure figure1 = new Bishop(true, from);
        figure1.figureMoved(Move.get(from, to));
        assertEquals(figure1.toString(), "Bishop-white-g5");
        figure1.figureMoved(Move.get(thirdpos, from));
        assertEquals(figure1.toString(), "Bishop-white-g5");
    }

    @Test
    public void testCastling() {
        Position from = Position.byCode("c1");
        Position to = Position.byCode("g5");
        Move move = Move.get(from, to);

        assertFalse(new Pawn(true, from).canCastle());

        Figure king = new King(true, from);
        assertTrue(king.canCastle(), "unmoved king");
        king.figureMoved(move);
        assertFalse(king.canCastle(), "moved king");

        Figure Rook = new Rook(false, from);
        assertTrue(Rook.canCastle(), "unmoved Rook");
        Rook.figureMoved(move);
        assertFalse(Rook.canCastle(), "moved Rook");
    }

    @Test
    public void testIsPassivetBound() {
        String des = "white 0 King-white-e1-0 Rook-white-e3-4 Queen-black-e5";
        SimpleArrayBoard game = new SimpleArrayBoard(des, mock(LastMoveProvider.class));

        Position from = Position.byCode("e3");
        Position to1 = Position.byCode("e2");
        Position to2 = Position.byCode("e4");
        Position to3 = Position.byCode("e5");
        Position to4 = Position.byCode("e6");
        Position to5 = Position.byCode("d3");

        Rook rook = new Rook(true, from);

        assertFalse(rook.isPassiveBound(to1, game));
        assertFalse(rook.isPassiveBound(to2, game));
        assertFalse(rook.isPassiveBound(to3, game));
        assertTrue(rook.isPassiveBound(to4, game));
        assertTrue(rook.isPassiveBound(to5, game));
    }

    @Test
    public void testIsBound() {
        String des = "white 0 King-white-e1-0 Rook-white-e3-4 Queen-black-e5";
        SimpleArrayBoard game = new SimpleArrayBoard(des, mock(LastMoveProvider.class));

        Position from1 = Position.byCode("e3");
        Position from2 = Position.byCode("e1");
        Position to1 = Position.byCode("e2");
        Position to2 = Position.byCode("e4");
        Position to3 = Position.byCode("e5");
        Position to4 = Position.byCode("e6");
        Position to5 = Position.byCode("d3");
        Position to6 = Position.byCode("d2");
        Position to7 = Position.byCode("a1");

        Rook rook = new Rook(true, from1);
        King king = new King(true, from2);

        assertFalse(rook.isBound(to1, game));
        assertFalse(rook.isBound(to2, game));
        assertFalse(rook.isBound(to3, game));
        assertTrue(rook.isBound(to4, game));
        assertTrue(rook.isBound(to5, game));
        assertFalse(king.isBound(to1, game));
        assertFalse(king.isBound(to6, game));

        des = "white 0 Rook-white-a1-0 King-white-e1-0 Rook-white-e3-4 Bishop-black-c3 Queen-black-e5";
        game.init(des);

        assertTrue(rook.isBound(to1, game));
        assertTrue(rook.isBound(to2, game));
        assertTrue(rook.isBound(to3, game));
        assertTrue(rook.isBound(to4, game));
        assertTrue(rook.isBound(to5, game));
        assertFalse(king.isBound(to1, game));
        assertTrue(king.isBound(to6, game));
        assertTrue(king.isBound(to7, game));

        des = "white 0 Rook-white-a1-0 King-white-e1-0 Bishop-black-e2";
        game.init(des);

        assertFalse(king.isBound(to1, game));
        assertFalse(king.isBound(to6, game));
        assertTrue(king.isBound(to7, game));
    }

    @Test
    public void testIsMoveable() {
        String des = "white 0 King-white-e1-0 Rook-white-h2-1 Queen-black-h4";
        SimpleArrayBoard game = new SimpleArrayBoard(des, mock(LastMoveProvider.class));

        Position from = Position.byCode("h2");
        Position to1 = Position.byCode("f2");
        Position to2 = Position.byCode("h4");
        Position to3 = Position.byCode("g3");
        Position to4 = Position.byCode("e2");

        Rook rook = new Rook(true, from);

        assertTrue(rook.isMovable(to1, game));
        assertTrue(rook.isMovable(to2, game));
        assertFalse(rook.isMovable(to3, game));
        assertFalse(rook.isMovable(to4, game));
    }

    @Test
    public void testGetTypeInfo() {
        Position from1 = Position.byCode("e3");
        Position from2 = Position.byCode("e1");

        Pawn pawn1 = new Pawn(true, from1);
        Pawn pawn2 = new Pawn(true, from1);
        Pawn pawn3 = new Pawn(false, from1);
        Pawn pawn4 = new Pawn(true, from2);

        assertEquals(pawn1.getTypeInfo(), pawn2.getTypeInfo());    //gleiche Objekte sollten die gleiche TypeInfo haben
        assertNotEquals(pawn1.getTypeInfo(), pawn3.getTypeInfo()); //unterschiedliche TypeInfo bei unterschiedliche Farbe
        assertEquals(pawn1.getTypeInfo(), pawn4.getTypeInfo());    //Position geht nicht mit ein

        Rook rook = new Rook(true, from1);
        Knight knight = new Knight(true, from1);
        Bishop bishop = new Bishop(true, from1);
        Queen queen = new Queen(true, from1);
        King king = new King(true, from2);

        Integer pawnByte = pawn1.getTypeInfo();
        Integer RookByte = rook.getTypeInfo();
        Integer knightByte = knight.getTypeInfo();
        Integer bishopByte = bishop.getTypeInfo();
        Integer queenByte = queen.getTypeInfo();
        Integer kingByte = king.getTypeInfo();

        List<Integer> figureByteList = new LinkedList<>();
        //Die byte-Werte müssen paarweise disjunkt sein
        figureByteList.add(pawnByte);
        assertFalse(
                figureByteList.contains(RookByte),
                "Bytewert sollte noch nicht in der Liste sein"
        );
        figureByteList.add(RookByte);
        assertFalse(
                figureByteList.contains(knightByte),
                "Bytewert sollte noch nicht in der Liste sein"
        );
        figureByteList.add(knightByte);
        assertFalse(
                figureByteList.contains(bishopByte),
                "Bytewert sollte noch nicht in der Liste sein"
        );
        figureByteList.add(bishopByte);
        assertFalse(
                figureByteList.contains(queenByte),
                "Bytewert sollte noch nicht in der Liste sein"
        );
        figureByteList.add(queenByte);
        assertFalse(
                figureByteList.contains(kingByte),
                "Bytewert sollte noch nicht in der Liste sein"
        );
    }

    static List<Move> getPossibleMovesFrom(String posCode, ChessGameInterface game) {
        Position from = Position.byCode(posCode);
        List<Move> moveIter = new LinkedList<>();
        List<Move> allMoves = new LinkedList<>();
        game.getPossibleMoves(allMoves);
        for (Move move : allMoves) {
            if (move.from.equalsPosition(from)) {
                moveIter.add(move);
            }
        }
        return moveIter;
    }
}

File: src/test/kotlin/voidchess/figures/FigureTest.kt
package voidchess.figures

import org.testng.Assert.*
import org.testng.annotations.DataProvider
import org.testng.annotations.Test
import voidchess.board.ChessGameInterface
import voidchess.board.getFigure
import voidchess.board.move.Move
import voidchess.board.move.Position
import voidchess.initSimpleChessBoard
import java.util.*


class FigureTest {

    @Test
    fun testIsDifferentColor() {
        val figure1 = Queen(true, Position[0, 0])
        val figure2 = Rook(true, Position[0, 0])
        val figure3 = Bishop(false, Position[0, 0])
        assertTrue(figure1.hasDifferentColor(figure3))
        assertFalse(figure2.hasDifferentColor(figure1))
        assertFalse(figure3.hasDifferentColor(figure3))
    }

    @Test
    fun testCanBeHitByEnpasent() {
        assertFalse(
                Knight(
                        true,
                        Position[0, 0]
                ).canBeHitByEnpasent(),
                "should be false for all voidchess.figures but pawns"
        )
    }

    @Test
    fun testToString() {
        val pos = Position.byCode("g4")
        val figure1 = Bishop(true, pos)
        assertEquals("Bishop-white-g4", figure1.toString())
        val figure2 = Bishop(false, pos)
        assertEquals("Bishop-black-g4", figure2.toString())
    }

    @Test
    fun testSubtypes() {
        val pos = Position.byCode("g4")
        Pawn(true, pos)
        Rook(true, pos)
        Knight(true, pos)
        Bishop(true, pos)
        Queen(true, pos)
        King(true, pos)
    }

    @Test
    fun testIsWhite() {
        val pos = Position.byCode("g4")

        val figure1 = Rook(true, pos)
        assertTrue(figure1.isWhite)

        val figure2 = Knight(false, pos)
        assertFalse(figure2.isWhite)
    }

    @Test
    fun testFigureMoved() {
        val from = Position.byCode("c1")
        val to = Position.byCode("g5")
        val thirdpos = Position.byCode("f5")

        val figure1 = Bishop(true, from)
        figure1.figureMoved(Move[from, to])
        assertEquals(figure1.toString(), "Bishop-white-g5")
        figure1.figureMoved(Move[thirdpos, from])
        assertEquals(figure1.toString(), "Bishop-white-g5")
    }

    @Test
    fun testCastling() {
        val from = Position.byCode("c1")
        val to = Position.byCode("g5")
        val move = Move[from, to]

        assertFalse(Pawn(true, from).canCastle())

        val king = King(true, from)
        assertTrue(king.canCastle(), "unmoved king")
        king.figureMoved(move)
        assertFalse(king.canCastle(), "moved king")

        val rook = Rook(false, from)
        assertTrue(rook.canCastle(), "unmoved Rook")
        rook.figureMoved(move)
        assertFalse(rook.canCastle(), "moved Rook")
    }

    @Test(dataProvider = "getIsBoundData")
    fun testIsBound(gameDes: String, figurePos: String, isNotBoundPosCodes: List<String>, isBoundPosCodes: List<String>) {
        val game = initSimpleChessBoard(gameDes)
        val figure = game.getFigure(Position.byCode(figurePos))
        for(isNotBoundPosCode in isNotBoundPosCodes) {
            val isNotBoundPos = Position.byCode(isNotBoundPosCode)
            assertFalse(figure.isBound(isNotBoundPos, game), "$figurePos.isBound($isNotBoundPos, game)")
        }
        for(isBoundPosCode in isBoundPosCodes) {
            val isBoundPos = Position.byCode(isBoundPosCode)
            assertTrue(figure.isBound(isBoundPos, game), "$figurePos.isBound($isBoundPos, game)")
        }
    }

    @DataProvider
    fun getIsBoundData(): Array<Array<Any>> = arrayOf(
            arrayOf("white 0 King-white-e1-0 Rook-white-e3-4 Queen-black-e5 King-black-e8-0", "e3", listOf("e2", "e4", "e5"), listOf("f3", "d3", "c3")),
            arrayOf("white 0 King-white-e1-0 Rook-white-e3-4 Queen-black-e5 King-black-e8-0", "e1", listOf("d1", "d2", "e2", "f2", "f1"), listOf<String>()),
            arrayOf("white 0 King-white-e1-0 Rook-white-e3-4 Bishop-black-c3 King-black-e8-0", "e3", listOf("c3"), listOf("d3", "f3", "e2", "e4", "e5")),
            arrayOf("white 0 King-white-e1-0 Rook-white-d3-4 Bishop-black-b4 King-black-e8-0", "d3", listOf("c3", "d2"), listOf("e3", "d4", "b3", "d1")),
            arrayOf("white 0 King-white-e1-0 Knight-white-a2 Bishop-black-b4 King-black-e8-0", "a2", listOf("c3", "b4"), listOf("c1")),
            arrayOf("white 0 Rook-white-a1-0 King-white-e1-0 Rook-white-e3-4 Bishop-black-c3 Queen-black-e5 King-black-e8-0", "e3", listOf<String>(), listOf("c3", "d3", "f3", "e2", "e4", "e5")),
            arrayOf("white 0 Rook-white-a1-0 King-white-e1-0 Rook-white-e3-4 Bishop-black-c3 Queen-black-e5 King-black-e8-0", "e1", listOf("d1", "e2", "f2", "f1"), listOf("a1", "d2")),
            arrayOf("white 0 Rook-white-a1-0 King-white-e1-0 Bishop-black-e2 King-black-e8-0", "e1", listOf("d2", "e2", "f2"), listOf("a1", "d1", "f1"))
    )

    @Test(dependsOnMethods = ["testIsBound"])
    fun testIsMovable() {
        val des = "white 0 King-white-e1-0 Rook-white-h2-1 Queen-black-h4 King-black-e8-0"
        val game = initSimpleChessBoard(des)

        val from = Position.byCode("h2")
        val to1 = Position.byCode("f2")
        val to2 = Position.byCode("h4")
        val to3 = Position.byCode("g3")
        val to4 = Position.byCode("e2")

        val rook = Rook(true, from)

        assertTrue(rook.isMovable(to1, game))
        assertTrue(rook.isMovable(to2, game))
        assertFalse(rook.isMovable(to3, game))
        assertFalse(rook.isMovable(to4, game))
    }

    @Test
    fun testGetTypeInfo() {
        val from1 = Position.byCode("e3")
        val from2 = Position.byCode("e1")

        val pawn1 = Pawn(true, from1)
        val pawn2 = Pawn(true, from1)
        val pawn3 = Pawn(false, from1)
        val pawn4 = Pawn(true, from2)

        assertEquals(pawn1.typeInfo, pawn2.typeInfo)    //gleiche Objekte sollten die gleiche TypeInfo haben
        assertNotEquals(pawn1.typeInfo, pawn3.typeInfo) //unterschiedliche TypeInfo bei unterschiedliche Farbe
        assertEquals(pawn1.typeInfo, pawn4.typeInfo)    //Position geht nicht mit ein

        val rook = Rook(true, from1)
        val knight = Knight(true, from1)
        val bishop = Bishop(true, from1)
        val queen = Queen(true, from1)
        val king = King(true, from2)

        val pawnByte = pawn1.typeInfo
        val rookByte = rook.typeInfo
        val knightByte = knight.typeInfo
        val bishopByte = bishop.typeInfo
        val queenByte = queen.typeInfo
        val kingByte = king.typeInfo

        val figureByteList = LinkedList<Int>()
        //Die byte-Werte müssen paarweise disjunkt sein
        figureByteList.add(pawnByte)
        assertFalse(
                figureByteList.contains(rookByte),
                "Bytewert sollte noch nicht in der Liste sein"
        )
        figureByteList.add(rookByte)
        assertFalse(
                figureByteList.contains(knightByte),
                "Bytewert sollte noch nicht in der Liste sein"
        )
        figureByteList.add(knightByte)
        assertFalse(
                figureByteList.contains(bishopByte),
                "Bytewert sollte noch nicht in der Liste sein"
        )
        figureByteList.add(bishopByte)
        assertFalse(
                figureByteList.contains(queenByte),
                "Bytewert sollte noch nicht in der Liste sein"
        )
        figureByteList.add(queenByte)
        assertFalse(
                figureByteList.contains(kingByte),
                "Bytewert sollte noch nicht in der Liste sein"
        )
    }
}
